### PR TITLE
Fix silent expert-weight transposition in MoE preprocess_weight

### LINKED ISSUE #849: MoE grouped_mm backend silently trains on transposed expert weights when `2*moe_intermediate == hidden_size` (bf16 and bnb-4bit)
`temporary_patches/moe_utils.py::preprocess_weight` converts transformers-v5 F.linear expert
layout (`gate_up (E, 2I, H)`, `down (E, H, I)`) to grouped_mm layout by shape inspection:

```python
    if proj_type == "gate_up":
        # Want (E, hidden_dim, 2*intermediate).
        if weight.shape[1] == hidden_dim:
            return weight
        else:
            return weight.transpose(-2, -1)
```

When `2*moe_intermediate_size == hidden_size` the check is ambiguous — `weight.shape[1]` equals
`hidden_dim` for the F.linear layout too — and the weight comes back **untransposed**, so the
gate_up matmul runs on transposed weights in every expert and every MoE layer (the analogous
`down` branch has the same ambiguity when `moe_intermediate_size == hidden_size`). Loss stays
finite and training appears to run normally, but outputs are numerically wrong. The
`_WEIGHT_PREPROCESSORS` registry that could disambiguate per-arch has no registrations
(2026.6.7 wheel and main as of 2026-07-02).

## Repro

unsloth 2026.6.9 / unsloth-zoo 2026.6.7 / transformers 5.5.0, torch 2.11.0+cu128 (outside
unsloth's `<2.11` pin — the layout logic is version-independent by code reading), any
grouped_mm-capable GPU:

```python
import unsloth, torch, torch.nn.functional as F
from transformers import Qwen3MoeConfig, Qwen3MoeForCausalLM

cfg = Qwen3MoeConfig(vocab_size=1024, hidden_size=64, intermediate_size=128,
                     moe_intermediate_size=32,   # 2*32 == hidden_size -> square gate_up
                     num_hidden_layers=1, num_attention_heads=4, num_key_value_heads=2,
                     head_dim=16, num_experts=8, num_experts_per_tok=2,
                     decoder_sparse_step=1, initializer_range=0.2)
torch.manual_seed(0)
model = Qwen3MoeForCausalLM(cfg).to("cuda", torch.bfloat16).eval()
experts = model.model.layers[0].mlp.experts

T, K = 64, 2
hs = torch.randn(T, cfg.hidden_size, device="cuda", dtype=torch.bfloat16)
idx = torch.randint(0, 8, (T, K), device="cuda")
w = torch.softmax(torch.rand(T, K, device="cuda", dtype=torch.bfloat16), -1)

out = experts(hs, idx, w).float()                      # zoo grouped_mm backend

ref = torch.zeros(T, cfg.hidden_size, device="cuda")   # stock Qwen3MoeExperts math, fp32
gu, dn = experts.gate_up_proj.float(), experts.down_proj.float()
for t in range(T):
    for k in range(K):
        e = idx[t, k]
        gate, up = F.linear(hs[t].float(), gu[e]).chunk(2, -1)
        ref[t] += w[t, k].float() * F.linear(F.silu(gate) * up, dn[e])

print("rel err:", ((out - ref).abs().max() / ref.abs().max()).item())
# grouped_mm backend (default):        rel err 0.994  (silently wrong)
# UNSLOTH_MOE_BACKEND=native_torch:    rel err 0.005  (bf16 noise, correct)
```

Measured on an RTX A2000: max-abs error vs stock math is **139× beyond** a bf16-precision control
on this bf16 repro, 140.7× on the `load_in_4bit` variant of the same model (the bnb-4bit dequant
path, `forward_moe_backend_bnb4bit`, feeds the same `preprocess_weight`, so 4-bit training is
equally affected). Forcing `preprocess_weight` to transpose-always drops the excess-over-noise
ratio to exactly 1.0 — max-abs error identical to the bf16-precision control — and
`torch._grouped_mm` itself is exact vs a hand loop, so the primitive is fine; the layout guess is
the bug. On pinned torch the same grouped_mm default path serves H100-class GPUs.

## Blast radius

As of today, no official checkpoint I scanned actually reaches the silent path: official
Qwen3-MoE, GLM-4.5, ERNIE-4.5, LFM2-8B-A1B and DeepSeek-V3 configs are all non-square, and
gpt-oss-20b/120b are down-square but safe because transformers stores gpt-oss already in grouped
layout, so no-transpose is correct there. But **allenai/OLMoE-1B-7B-0924 (and -0125) has square
gate_up** (2·1024 == 2048) — it currently crashes earlier for an unrelated reason (#850), and the obvious fix for that crash routes it straight into this silent one. Any
pruned/community config with matching dims hits it silently now.

## Suggested fix

The layout is knowable from the source module class — register per-arch preprocessors (the
registry already exists, unused), or key on a `_unsloth_grouped_mm_format` attribute the way zoo
PR #717 did for `GptOssExperts`. Shape-guessing can't be made safe for square dims.

Workaround for affected users: `UNSLOTH_MOE_BACKEND=native_torch`.

Found while verifying the #4032 fix (which works for its Qwen3-30B-A3B target in my testing) —
happy to test candidate patches; the repro and parity harness are ready to rerun.

### PR DESCRIPTION
**Pull request by BardiaKoopah**
# Fix silent expert-weight transposition in MoE `preprocess_weight`

## Summary

- Disambiguate square expert weights using the guaranteed non-square sibling projection.
- Pass the experts module through grouped MM and FP8 preprocessing paths.
- Resolve PEFT target parameters with `get_param()` before walking `base_layer` wrappers.
- Read bitsandbytes logical `_original_shape` after unwrapping a linear module's `.weight`.
- Warn and default to the transformers `F.linear` layout when no readable sibling exists.

## Root cause

Shape-only inference cannot distinguish `F.linear` and grouped MM layouts when an expert weight is square. The first shape helper also inspected wrapper objects before reaching the underlying PEFT or bitsandbytes parameter, so valid sibling layout information could be lost.

Within a fused experts module, `gate_up` and `down` cannot both be square. The non-square sibling therefore identifies the shared storage layout without dequantizing the weights.

## Behavior

- Non-square weights keep the existing fast path.
- Square `gate_up` and `down` weights use the sibling projection to select the correct orientation.
- PEFT target-parameter wrappers and bitsandbytes `Linear4bit` containers preserve their logical shapes.
- Model-specific preprocessors still take precedence.

## Validation

- `tests/test_moe_preprocess_weight_transpose.py`: 8 passed.
- Real PEFT 0.18 `ParamWrapper` shape resolution passed.
- Python compile check passed.
- Targeted ruff check passed.
- The reporter's CUDA harness confirmed parity for square `gate_up`, square `down`, 4-bit, and non-square cases.

Thanks to @pjordanandrsn for the diagnosis and CUDA validation.

Fixes #849

### CODE REVIEW

**Comment by BardiaKoopah** [comment]
@pjordanandrsn check this out. Should fix your bug

**Comment by pjordanandrsn** [comment]
Closed the CUDA loop — re-ran the #849 harness on this branch, same setup as the original report (RTX A2000, torch 2.11.0+cu128, transformers 5.5.0, bitsandbytes 0.49.2, unsloth 2026.6.9), zoo swapped between main @ 25b7c07 and this branch @ a88cf17. Verdict: **fixed everywhere I can measure, no regression on the non-square fast path.**

| case | main @ 25b7c07 | this branch @ a88cf17 |
|---|---|---|
| #849 gist (square gate_up, bf16, default backend) — rel err | 0.994 | **0.0053** — identical to the `native_torch` control to every printed digit |
| tiny square gate_up, **bnb-4bit** — excess over bf16-noise | 140.7× | **1.0** |
| tiny square **down** (`I == H`), bf16 — excess | 181.4× | **1.0** |
| tiny square down, bnb-4bit — excess | 284.3× | **1.0** |
| non-square control, bf16 + 4-bit — excess | 1.0 | **1.0**, max-abs error bit-identical to main's |
| `tests/test_moe_preprocess_weight_transpose.py` | — | 7/7 |

Notes:

- The square-**down** rows are a case my original repro never drove (it only exercised square gate_up). On main the down branch is broken the same way (181× beyond the bf16-noise floor); on this branch the non-square gate_up sibling resolves it, exactly as the pair-can't-both-be-square argument predicts.
- "Excess 1.0" here is the strong sense from the original report: max-abs error vs the fp32 reference comes out float-identical to the bf16-precision control.
- Reading `_original_shape` off the bnb `Params4bit` so the sibling vote never dequantizes is a nice touch — the 4-bit path stays cheap.
- One code-read observation (not measured — I have no model on this path): the sibling vote also resolves modules stored pre-grouped (`gpt-oss`-style layouts) correctly, which the old down-branch only got right by coincidence of its shape test.
- Scope matches the earlier thread findings: this single choke point covers grouped_mm, bnb-4bit, and both FP8 routes (FP8 remains code-read only on my side — no Hopper hardware); `unsloth_triton` was measured unaffected for the square case upthread.

Thanks @BardiaKoopah — from the reporter's side this fixes #849. Happy to re-run anything if the branch changes.

**Comment by shimmyshimmer** [comment]
@codex

File: unsloth_zoo/temporary_patches/moe_utils.py [hunk 1]
```diff
@@ -886,31 +886,122 @@ def get_weight_preprocessor(model_type: str):
...
+    while hasattr(param, "base_layer"):
+        param = param.base_layer
+    original_shape = getattr(param, "_original_shape", None)
+    if original_shape is not None:
+        return tuple(int(x) for x in original_shape)
+    # Plain Parameters / Linear layers expose the logical shape directly; only unwrap a
+    # `.weight` container (a Linear-like module), not a `get_param` provider we'd rather
+    # not materialize.
+    if not hasattr(param, "get_param") and hasattr(param, "weight"):
+        param = param.weight
```
[the remark this answers is not in the document]
Comment (shimmyshimmer) [reply unsloth_zoo/temporary_patches/moe_utils.py]: Fixed in 7dfc866b. `_logical_expert_shape` now unwraps linear `.weight` containers before reading `_original_shape`, with regression coverage for a packed `Params4bit` shape.

File: unsloth_zoo/temporary_patches/moe_utils.py [hunk 2]
```diff
@@ -886,31 +886,122 @@ def get_weight_preprocessor(model_type: str):
...
+    """Best-effort logical (E, d1, d2) shape of an expert weight without dequantizing.
+
+    Mirrors _get_base_weight's unwrap order but reads shape only. A bnb Params4bit's
+    .shape is the packed uint8 layout, so prefer its recorded _original_shape. Returns
+    None if the logical shape can't be determined cheaply (caller degrades gracefully).
+    """
+    # This Unsloth Zoo code section is licensed under AGPL3
+
+    while hasattr(param, "base_layer"):
+        param = param.base_layer
```
[the remark this answers is not in the document]
Comment (shimmyshimmer) [reply unsloth_zoo/temporary_patches/moe_utils.py]: Fixed in 7dfc866b. `_logical_expert_shape` now resolves `get_param()` before `base_layer`, preserving grouped square weights. Added regression coverage for a PEFT-style wrapped sibling.

**Comment by shimmyshimmer** [comment]
@codex

**Comment by danielhanchen** [comment]
Follow-up on top of the wrapped-sibling resolution fix.

`_logical_expert_shape` now reads a recorded logical shape (`Params4bit._original_shape` or `ParameterModule.shape_3d`) before falling back to `get_param()`, so resolving a wrapped square sibling no longer materializes a full contiguous expert stack just to read its shape. No behavior change for any case the existing tests cover; added `test_logical_expert_shape_reads_shape_3d_without_materializing` to pin it.

Also verified end to end: on a B200 a grouped_mm expert forward reproduces #849 on the pre-fix base (square gate_up at 358x the bf16 noise floor, square down at 387x) and both square variants return to the noise floor on this branch, with the non-square control bit-identical to base. The CPU regression suite is 9/9 and the adjacent grouped_mm suites stay green.
